hyperspy/model.py

```diff
@@ -891,6 +898,312 @@ def _model_function(self, param):
+            warnings.warn(
+                "The model contains custom components not based on Expression, with more than"
+                "one component. "
+                "This is not recommended for linear optimizers, since the optimizer is unable "
+                "to determine if there are any parts of the component that are constant. "
+                "Example: comp = a*x+b, with b.free = False and non-zero. "
+                "Either continue at own risk, or define the component as a `Expression` instead."
+                )
+
+        self._set_linear_parameters_to_one()
+        self._set_p0()
+
+        channels_signal_shape = np.count_nonzero(self.channel_switches)
+
+        self._component_data = np.zeros((n_free_para, channels_signal_shape))
+        self._component_data_fixed = np.zeros(channels_signal_shape)
+
+        self._set_twinned_lists()
+        for component in self:
+            # only compute if component and topmost parent are active
+            if component.active and check_top_parent_twins_are_active(component):
+                self._compute_component(component)
+
+        if self._linear_ndfit:
+            nav_shape = self.axes_manager._navigation_shape_in_array
+            if self.channel_switches.all():
+                # If channel_switches is all True, then is much more performant
+                # and less memory-intensive to just reshape the signal than use 
+                # fancy indexing (which would create a copy of the signal),
+                # especially with dask
+                target_signal = self.signal.data.reshape(nav_shape + (-1,))
+            else:
+                target_signal = self.signal.data.T[np.where(self.channel_switches.T)].T
+        else:
+            target_signal = self.signal()[np.where(self.channel_switches)]
+            nav_shape = ()
+
+        if is_binned(self.signal):
+            target_signal = target_signal / np.prod(
+                tuple((ax.scale for ax in self.signal.axes_manager.signal_axes))
+            )
+
+        target_signal = target_signal - self._component_data_fixed
+
+        flat_sig_len = target_signal.shape[-1] #np.count_nonzero(self.channel_switches)
+
+        # Reshape what may potentially be Signal2D data into a long Signal1D shape
+        # and an nD navigation shape to a 1D nav shape
+        if self._linear_ndfit:
+            target_signal = target_signal.reshape((np.prod(nav_shape, dtype=int), ) + (flat_sig_len,))
+        else:
+            target_signal = target_signal.reshape((flat_sig_len,))
+
+        if optimizer == "lstsq" and not self.signal._lazy:
+            result, residual, *_ = np.linalg.lstsq(self._component_data.T, target_signal.T, rcond=None)
+            self.coefficient_array = result.T
+
+        elif optimizer == "lstsq" and self.signal._lazy:
+            result, residual, *_ = da.linalg.lstsq(da.asarray(self._component_data).T, target_signal.T)
```

Comment:
So, given that `self.signal._lazy` is True, `target_signal` is a dask array. `self._component_data` will be a numpy array, since there is (currently?) no point in calculating the components lazily. `da.linalg.lstsq` expects a dask array as its first and second argument and will raise an error if the first argument is a numpy array, see this example:

```python
da.linalg.lstsq(np.random.random((100,10)), da.random.random((100, 5)))
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-4-6b74bf26dfa9> in <module>
----> 1 da.linalg.lstsq(np.random.random((100,10)), da.random.random((100, 5)))

~/mambaforge/envs/hspy/lib/python3.9/site-packages/dask/array/linalg.py in lstsq(a, b)
   1403         Singular values of `a`.
   1404     """
-> 1405     q, r = qr(a)
   1406     x = solve_triangular(r, q.T.conj().dot(b))
   1407     residuals = b - a.dot(x)

~/mambaforge/envs/hspy/lib/python3.9/site-packages/dask/array/linalg.py in qr(a)
    857     """
    858
--> 859     if len(a.chunks[1]) == 1 and len(a.chunks[0]) > 1:
    860         return tsqr(a)
    861     elif len(a.chunks[0]) == 1:

AttributeError: 'numpy.ndarray' object has no attribute 'chunks'
```

The other way round, `da.linalg.lstsq(da.random.random((100,10)), np.random.random((100, 5)))`, works.